elf/input-files.cc

```diff
@@ -1314,20 +1314,41 @@ std::ostream &operator<<(std::ostream &out, const InputFile<E> &file) {
 template <typename E>
 MappedFile<Context<E>> *find_dynamic_library(Context<E> &ctx,
                                              std::string lib,
-                                             const std::vector<std::string_view> &runpath) {
+                                             const std::vector<std::string_view> &runpath,
+                                             bool runpath_first) {
   // `lib` here is something like `libfoo.so`
   auto try_open = [&ctx](std::string_view dir, const std::string &lib) {
     std::string path = std::string(dir) + "/" + lib;
     return open_library(ctx, path);
   };
 
-  for (std::string_view dir : runpath)
-    if (MappedFile<Context<E>> *mf = try_open(dir, lib))
-      return mf;
+  auto search_library_paths = [&ctx, &lib, &try_open]() {
+    MappedFile<Context<E>> *mf;
+    for (std::string_view dir : ctx.arg.library_paths)
+      if (mf = try_open(dir, lib))
+        return mf;
+    return mf;
+  };
 
-  for (std::string_view dir : ctx.arg.library_paths)
-    if (MappedFile<Context<E>> *mf = try_open(dir, lib))
+  auto search_runpath = [&ctx, &lib, &runpath, &try_open]() {
+    MappedFile<Context<E>> *mf;
+    for (std::string_view dir : runpath)
+      if (mf = try_open(dir, lib))
+        return mf;
+    return mf;
+  };
+
+  if (runpath_first) {
+    if (MappedFile<Context<E>> *mf = search_runpath())
+      return mf;
+    else if (MappedFile<Context<E>> *mf = search_library_paths())
       return mf;
+  } else {
+    if (MappedFile<Context<E>> *mf = search_library_paths())
+      return mf;
+    else if (MappedFile<Context<E>> *mf = search_runpath())
+      return mf;
+  }
```

Comment:
I'm not exactly sure how to make this prettier. I would like to chain the iterators in a different order based on the value of `runpath_first`, but I couldn't find a way to do it with the ranges library... Is there a better way?